proxy/http2/Http2DependencyTree.h

```diff
@@ -155,23 +164,41 @@ Http2DependencyTree<T>::_find(Node *node, uint32_t id, uint32_t depth)
 }
 
 template <typename T>
-typename Http2DependencyTree<T>::Node *
-Http2DependencyTree<T>::find(uint32_t id)
+Node *
+Tree<T>::find_shadow(uint32_t id)
 {
   return _find(_root, id);
 }
 
 template <typename T>
-typename Http2DependencyTree<T>::Node *
-Http2DependencyTree<T>::add(uint32_t parent_id, uint32_t id, uint32_t weight, bool exclusive, T t)
+Node *
+Tree<T>::find(uint32_t id)
+{
+  Node *n = _find(_root, id);
+  return n == nullptr ? nullptr : (n->is_shadow() ? nullptr : n);
+}
+
+template <typename T>
+Node *
+Tree<T>::add(uint32_t parent_id, uint32_t id, uint32_t weight, bool exclusive, T t)
 {
   Node *parent = find(parent_id);
-  if (parent == NULL) {
-    parent = _root;
+  if (parent == nullptr) {
+    if (exclusive) {
+      parent = add(0, parent_id, weight, false, nullptr);
+    } else {
+      parent = _root;
+    }
+  }
+
+  Node *node = find(id);
+  if (node != NULL && node->is_shadow()) {
```

Comment:
It looks like `find(id)` never returns shadow node, right? If so, this condition never be true.
Do you want to call `find_shadow(id)` ?